pyenzyme/tools.py

```diff
@@ -149,6 +151,199 @@ def extract(obj, target) -> list:
     return result
 
 
+def group_measurements(
+    enzymemldoc: EnzymeMLDocument,
+    attribute: Literal["initial", "prepared"] = "initial",
+    tolerance: float = 0.0,
+) -> EnzymeMLDocument:
+    """
+    Groups measurements by their conditions and assigns for each unique set of conditions
+    a `group_id` attribute.
+
+    Args:
+        enzymemldoc (EnzymeMLDocument): The EnzymeMLDocument to group measurements from.
+        attribute (Literal["initial", "prepared"]): Which concentration attribute to use for grouping.
+        tolerance (float): Percentage tolerance for numerical comparison (e.g., 0.05 for 5%) for matching groups.
+
+    Returns:
+        EnzymeMLDocument: The EnzymeMLDocument with grouped measurements.
+    """
+
+    # Create a mapping from conditions to group_id
+    conditions_to_group: dict[tuple, str] = {}
+    group_counter = 0
+
+    for measurement in enzymemldoc.measurements:
+        # Get conditions as a hashable representation
+        conditions = _get_measurement_conditions(measurement, attribute)
+
+        # Find matching group with tolerance
+        matching_group = _find_matching_group(
+            conditions, conditions_to_group, tolerance
+        )
+
+        if matching_group is None:
+            # Create new group
+            group_id = f"group_{group_counter}"
+            conditions_hash = _conditions_to_hashable(conditions)
+            conditions_to_group[conditions_hash] = group_id
+            group_counter += 1
+            measurement.group_id = group_id
+        else:
+            measurement.group_id = matching_group
+
+    return enzymemldoc
+
+
+def _get_measurement_conditions(
+    measurement: Measurement,
+    attribute: Literal["initial", "prepared"] = "initial",
+) -> dict:
+    """
+    Extract measurement conditions as a dictionary, inclding pH and temperature.
+    Does not account for variation in units.
+
+    Args:
+        measurement: The measurement to extract conditions from
+        attribute: Which concentration attribute to use ("initial" or "prepared")
+
+    Returns:
+        dict: Conditions dictionary with species concentrations and environmental conditions
+    """
+    conditions = {}
+
+    # Add species conditions using the specified attribute
+    for species_data in measurement.species_data:
+        if attribute == "initial":
```

Comment:
You can use `getattr` to retrieve any of these without needing to branch, but this is fine.